src/xml_utils.cpp

```diff
@@ -0,0 +1,1850 @@
+
+  do {
+    zs.next_out = reinterpret_cast<Bytef*>(outbuffer);
+    zs.avail_out = sizeof(outbuffer);
+
+    ret = deflate(&zs, Z_FINISH);
+
+    if (compressed_data.size() < zs.total_out) {
+      compressed_data.insert(compressed_data.end(), outbuffer, outbuffer + (zs.total_out - compressed_data.size()));
+    }
+
+  } while (ret == Z_OK);
+
+  deflateEnd(&zs);
+
+  return std::string(compressed_data.begin(), compressed_data.end());
+}
+
+
+
+
+std::string xml_utils::decompress_zlib(const std::string& compressed_string) {
+
+  // std::string output_string;
+  //
+  // z_stream zs;
+  // memset(&zs, 0, sizeof(zs));
+  //
+  // inflateInit(&zs);
+  //
+  // // if (inflateInit(&zs) != Z_OK) {
+  // //   throw std::runtime_error("inflateInit failed!");
+  // // }
+  //
+  // zs.next_in = (Bytef*)compressed_string.data();
+  // zs.avail_in = compressed_string.size();
+  //
+  // int ret;
+  // char outbuffer[32768];
+  //
+  // do {
+  //   zs.next_out = reinterpret_cast<Bytef*>(outbuffer);
+  //   zs.avail_out = sizeof(outbuffer);
+  //
+  //   ret = inflate(&zs, 0);
+  //
+  //   if (output_string.size() < zs.total_out) {
+  //     output_string.append(outbuffer, zs.total_out - output_string.size());
+  //   }
+  //
+  // } while (ret == Z_OK);
+  //
+  // inflateEnd(&zs);
+  //
+  // // if (inflateEnd(&zs) != Z_OK) {
+  // //   throw std::runtime_error("inflateEnd failed!");
+  // // }
+  //
+  // return output_string;
+
+  z_stream zs;
+  memset(&zs, 0, sizeof(zs));
+  inflateInit(&zs);
+  zs.next_in = (Bytef*)compressed_string.data();
+  zs.avail_in = compressed_string.size();
+
+  int ret;
+  char outbuffer[32768];
+  std::string outstring;
+  do
+  {
+    zs.next_out = reinterpret_cast<Bytef*>(outbuffer);
+    zs.avail_out = sizeof(outbuffer);
+    ret = inflate(&zs, 0);
+    if (outstring.size() < zs.total_out)
+    {
+      outstring.append(outbuffer, zs.total_out - outstring.size());
+    }
+  }
+  while (ret == Z_OK);
+  inflateEnd(&zs);
+
+  return outstring;
+}
+
+
+
+
+std::string xml_utils::encode_little_endian(const Rcpp::NumericVector& input) {
+
+  size_t n = input.size();
+
+  std::string result(n * sizeof(double), '\0');
+
+  const uint8_t* bytes = reinterpret_cast<const uint8_t*>(input.begin());
+
+  uint8_t* output_bytes = reinterpret_cast<uint8_t*>(&result[0]);
+
+  for (size_t i = 0; i < n; ++i) {
+    uint64_t little_endian_value;
+    std::memcpy(&little_endian_value, bytes + i * sizeof(double), sizeof(double));
+    std::memcpy(output_bytes + i * sizeof(uint64_t), &little_endian_value, sizeof(uint64_t));
+  }
+
+  return result;
+}
+
+
+
+
+Rcpp::NumericVector xml_utils::decode_little_endian(const std::string& str, const int& precision) {
+
+  std::vector<unsigned char> byteVec(str.begin(), str.end());
+
+  int array_len = (byteVec.size() / precision);
+
+  Rcpp::NumericVector result(array_len);
+
+  for (int i = 0; i < array_len; ++i) {
+    result[i] = reinterpret_cast<double&>(byteVec[i * precision]);
```

Comment:
This doesn't work for float data. I guess this should be a template function instead? Where double/float/int is used alongside the precision.